File: lib/models/game_state.dart
import 'game.dart';

enum CurrentPlayer { PLAYER1, PLAYER2 }

class GameState {
  final int id;
  final Game game;
  final int player1Score;
  final int player2Score;
  final int player1Legs;
  final int player2Legs;
  final int player1Sets;
  final int player2Sets;
  final CurrentPlayer currentPlayer;
  final int roundNumber;
  final int player1DartsThrown;
  final int player2DartsThrown;
  final DateTime updatedAt;

  GameState({
    required this.id,
    required this.game,
    required this.player1Score,
    required this.player2Score,
    this.player1Legs = 0,
    this.player2Legs = 0,
    this.player1Sets = 0,
    this.player2Sets = 0,
    this.currentPlayer = CurrentPlayer.PLAYER1,
    this.roundNumber = 1,
    this.player1DartsThrown = 0,
    this.player2DartsThrown = 0,
    required this.updatedAt,
  });

  factory GameState.fromJson(Map<String, dynamic> json) {
    return GameState(
      id: json['id'] as int,
      game: Game.fromJson(json['game']),
      player1Score: json['player1Score'] as int,
      player2Score: json['player2Score'] as int,
      player1Legs: json['player1Legs'] as int? ?? 0,
      player2Legs: json['player2Legs'] as int? ?? 0,
      player1Sets: json['player1Sets'] as int? ?? 0,
      player2Sets: json['player2Sets'] as int? ?? 0,
      currentPlayer: CurrentPlayer.values.firstWhere(
        (e) => e.toString() == 'CurrentPlayer.${json['currentPlayer']}',
      ),
      roundNumber: json['roundNumber'] as int? ?? 1,
      player1DartsThrown: json['player1DartsThrown'] as int? ?? 0,
      player2DartsThrown: json['player2DartsThrown'] as int? ?? 0,
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'game': game.toJson(),
      'player1Score': player1Score,
      'player2Score': player2Score,
      'player1Legs': player1Legs,
      'player2Legs': player2Legs,
      'player1Sets': player1Sets,
      'player2Sets': player2Sets,
      'currentPlayer': currentPlayer.toString().split('.').last,
      'roundNumber': roundNumber,
      'player1DartsThrown': player1DartsThrown,
      'player2DartsThrown': player2DartsThrown,
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  GameState copyWith({
    int? id,
    Game? game,
    int? player1Score,
    int? player2Score,
    int? player1Legs,
    int? player2Legs,
    int? player1Sets,
    int? player2Sets,
    CurrentPlayer? currentPlayer,
    int? roundNumber,
    int? player1DartsThrown,
    int? player2DartsThrown,
    DateTime? updatedAt,
  }) {
    return GameState(
      id: id ?? this.id,
      game: game ?? this.game,
      player1Score: player1Score ?? this.player1Score,
      player2Score: player2Score ?? this.player2Score,
      player1Legs: player1Legs ?? this.player1Legs,
      player2Legs: player2Legs ?? this.player2Legs,
      player1Sets: player1Sets ?? this.player1Sets,
      player2Sets: player2Sets ?? this.player2Sets,
      currentPlayer: currentPlayer ?? this.currentPlayer,
      roundNumber: roundNumber ?? this.roundNumber,
      player1DartsThrown: player1DartsThrown ?? this.player1DartsThrown,
      player2DartsThrown: player2DartsThrown ?? this.player2DartsThrown,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'GameState{id: $id, player1Score: $player1Score, player2Score: $player2Score, currentPlayer: $currentPlayer}';
  }
}

File: backend/src/main/java/com/dartverein/model/GameState.java
package com.dartverein.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
@Table(name = "game_state")
public class GameState {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "game_id", nullable = false, unique = true)
    @NotNull
    private Game game;

    @Column(name = "player1_score", nullable = false)
    @NotNull
    private Integer player1Score;

    @Column(name = "player2_score", nullable = false)
    @NotNull
    private Integer player2Score;

    @Column(name = "player1_legs")
    private Integer player1Legs = 0;

    @Column(name = "player2_legs")
    private Integer player2Legs = 0;

    @Column(name = "player1_sets")
    private Integer player1Sets = 0;

    @Column(name = "player2_sets")
    private Integer player2Sets = 0;

    @Enumerated(EnumType.STRING)
    @Column(name = "current_player", nullable = false, length = 20)
    @NotNull
    private CurrentPlayer currentPlayer;

    @Column(name = "round_number")
    private Integer roundNumber = 1;

    @Column(name = "player1_darts_thrown")
    private Integer player1DartsThrown = 0;

    @Column(name = "player2_darts_thrown")
    private Integer player2DartsThrown = 0;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum CurrentPlayer {
        PLAYER1, PLAYER2
    }

    // Constructors
    public GameState() {
        this.updatedAt = LocalDateTime.now();
    }

    public GameState(Game game, Integer startScore) {
        this();
        this.game = game;
        this.player1Score = startScore;
        this.player2Score = startScore;
        this.currentPlayer = CurrentPlayer.PLAYER1;
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Game getGame() {
        return game;
    }

    public void setGame(Game game) {
        this.game = game;
    }

    public Integer getPlayer1Score() {
        return player1Score;
    }

    public void setPlayer1Score(Integer player1Score) {
        this.player1Score = player1Score;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer2Score() {
        return player2Score;
    }

    public void setPlayer2Score(Integer player2Score) {
        this.player2Score = player2Score;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer1Legs() {
        return player1Legs;
    }

    public void setPlayer1Legs(Integer player1Legs) {
        this.player1Legs = player1Legs;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer2Legs() {
        return player2Legs;
    }

    public void setPlayer2Legs(Integer player2Legs) {
        this.player2Legs = player2Legs;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer1Sets() {
        return player1Sets;
    }

    public void setPlayer1Sets(Integer player1Sets) {
        this.player1Sets = player1Sets;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer2Sets() {
        return player2Sets;
    }

    public void setPlayer2Sets(Integer player2Sets) {
        this.player2Sets = player2Sets;
        this.updatedAt = LocalDateTime.now();
    }

    public CurrentPlayer getCurrentPlayer() {
        return currentPlayer;
    }

    public void setCurrentPlayer(CurrentPlayer currentPlayer) {
        this.currentPlayer = currentPlayer;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getRoundNumber() {
        return roundNumber;
    }

    public void setRoundNumber(Integer roundNumber) {
        this.roundNumber = roundNumber;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer1DartsThrown() {
        return player1DartsThrown;
    }

    public void setPlayer1DartsThrown(Integer player1DartsThrown) {
        this.player1DartsThrown = player1DartsThrown;
        this.updatedAt = LocalDateTime.now();
    }

    public Integer getPlayer2DartsThrown() {
        return player2DartsThrown;
    }

    public void setPlayer2DartsThrown(Integer player2DartsThrown) {
        this.player2DartsThrown = player2DartsThrown;
        this.updatedAt = LocalDateTime.now();
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}